rdar://89218473 ([ER] WebKitTestRunner and DumpRenderTree should let the user know if loading a file or a url failed)

mscottapple opened the pull request:
#### b3b218402da7205aec197cdc863e9d3f63fc3b95
<pre>
rdar://89218473 ([ER] WebKitTestRunner and DumpRenderTree should let the user know if loading a file or a url failed)

Reviewed by David Kilzer (ddkilzer).

This adds preflight checking for the existence of input files specified on the command line, as well as reporting of any URL loading errors.

* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(initializeGlobalsFromCommandLineOptions):
(runTest):
* Tools/DumpRenderTree/mac/ResourceLoadDelegate.h:
* Tools/DumpRenderTree/mac/ResourceLoadDelegate.mm:
(-[ResourceLoadDelegate dealloc]):
(-[ResourceLoadDelegate webView:resource:didFailLoadingWithError:fromDataSource:]):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::createOtherPlatformWebView):
(WTR::TestController::createWebViewWithOptions):
(WTR::TestController::createTestURL):
(WTR::TestController::runTest):
(WTR::TestController::didFailProvisionalNavigation):
(WTR::pathSuitableForTestResult):
* Tools/WebKitTestRunner/TestController.h:

Canonical link: <a href="https://commits.webkit.org/268687@main">https://commits.webkit.org/268687@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9b7081a1ca1f20f2ff3d8bc91636d74e59097937

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20294 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20726 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21367 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22194 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18938 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20526 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23983 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/20897 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20514 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20411 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17646 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23044 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17580 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18449 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24701 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18656 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18625 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22674 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19210 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16311 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18415 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4895 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22756 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19035 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Tools/DumpRenderTree/mac/DumpRenderTree.mm:
```diff
@@ -1093,7 +1098,7 @@ static void runTestingServerLoop()
             continue;
 
-        runTest(filenameBuffer);
+        runTest(filenameBuffer, true);
```
Try to avoid the boolean trap here by adding an enum.

smfr commented on Tools/WebKitTestRunner/TestController.cpp:
```diff
@@ -1464,17 +1464,26 @@ WKRetainPtr<WKStringRef> TestController::backgroundFetchState(WKStringRef)
         return 0;
 
+    if (length >= 7 && strstr(pathOrURL, "file://")) {
+        if (!WTF::FileSystemImpl::fileExists(String(pathOrURL + 7, length - 7))) {
```
How much does this new `stat` slow down `run-webkit-tests`?

ddkilzer replied:
Can we avoid this code path in "server" mode?

ddkilzer commented on Tools/DumpRenderTree/mac/DumpRenderTree.mm:
```diff
@@ -1895,7 +1900,7 @@ static void WebThreadLockAfterDelegateCallbacksHaveCompleted()
 }
 
-static void runTest(const std::string& inputLine)
+static void runTest(const std::string& inputLine, bool isServerMode)
```
Please add an enum instead of a `bool` here.  Maybe something like:

```
enum class TestingMode : bool { LocalMode, ServerMode };
```

ddkilzer reviewed (changes requested):
Let's add an enum as Simon suggested, and can we avoid the `stat` if we're in `ServerMode`?

ddkilzer commented on Tools/DumpRenderTree/mac/DumpRenderTree.mm:
```diff
@@ -1915,6 +1922,19 @@ static void runTest(const std::string& inputLine)
+        }
+    }
+
+    resourceLoadDelegate().get().mainResourceURL = (gUsingServerMode ? nullptr : [url _webkit_canonicalize_with_wtf]);
```
I think `mainResourceURL` will never be set (will always be `nil`) if `gUsingServerMode` is true, so it would probably be a bit easier to read like this (also, since this is an ObjC object, `nil` is preferred to `nullptr`, though this change eliminates that):

```suggestion
    if (!gUsingServerMode)
        resourceLoadDelegate().get().mainResourceURL = [url _webkit_canonicalize_with_wtf]);
```

This could also be combined with the `if` block above.

ddkilzer commented on Tools/DumpRenderTree/mac/ResourceLoadDelegate.mm:
```diff
@@ -263,7 +272,13 @@ -(void)webView: (WebView *)wv resource:identifier didFinishLoadingFromDataSource
 -(void)webView: (WebView *)wv resource:identifier didFailLoadingWithError:(NSError *)error fromDataSource:(WebDataSource *)dataSource
 {
-    if (!done && gTestRunner->dumpResourceLoadCallbacks()) {
+    if (done) {
```
This code path should probably only be active if `!gUsingServerMode` is true, correct?

```suggestion
    if (!gUsingServerMode && done) {
```

mscottapple replied:
I was relying on `mainResourceURL` being `nil` but now that I've got a global variable accessible here, this is better. Thanks.

ddkilzer reviewed (approved):
r=me, but a couple minor fixes requested.

mscottapple committed:
rdar://89218473 ([ER] WebKitTestRunner and DumpRenderTree should let the user know if loading a file or a url failed)

Reviewed by David Kilzer (ddkilzer).

This adds preflight checking for the existence of input files specified on the command line, as well as reporting of any URL loading errors.

* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(initializeGlobalsFromCommandLineOptions):
(runTest):
* Tools/DumpRenderTree/mac/ResourceLoadDelegate.h:
* Tools/DumpRenderTree/mac/ResourceLoadDelegate.mm:
(-[ResourceLoadDelegate dealloc]):
(-[ResourceLoadDelegate webView:resource:didFailLoadingWithError:fromDataSource:]):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::createOtherPlatformWebView):
(WTR::TestController::createWebViewWithOptions):
(WTR::TestController::createTestURL):
(WTR::TestController::runTest):
(WTR::TestController::didFailProvisionalNavigation):
(WTR::pathSuitableForTestResult):
* Tools/WebKitTestRunner/TestController.h:

Canonical link: https://commits.webkit.org/268687@main
